service/controller/release/resource/label/create.go

```diff
@@ -0,0 +1,70 @@
+package status
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToReleaseCR(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding if custom resource needs to be updated")
+
+		currentLabels := cr.Labels
+		if currentLabels == nil {
+			currentLabels = map[string]string{}
+		}
```

Comment:
That does not seem to make sense. What is `currentLabels` good for?

Reply:
The code below got updated. They are here to make sure we not try to get value from `nil` map.

Reply:
I am still wondering why we use a new variable instead of simply `cr.Labels`. That is even shorter than the variable you chose and the value is a map anyway so copying it does not do something different.

Reply:
Are you sure `cr.Labels` can't be null? I don't want to mutate the CR here.

Reply:
It can, but why not just this? 

```
		if cr.Labels == nil {
			cr.Labels = map[string]string{}
		}
```

Reply:
It will change the `cr` for other resources as well. Also I think the object from informer cache should not be mutated. I remember I saw that in some godoc.

Every time we update the object or its status we fetch the new instance directly from the API. I think the watched object shouldn't be mutated.

Reply:
Working on the variable changes the CR as well. Because it is a map. That is a pointer. You do not copy it with a new CR. And then, we could just copy the CR and work with it in the resource. We could even only give a copy to all resources in `operatorkit` by default.

Reply:
Map is a value that can be `nil`. It isn't a pointer. See https://play.golang.org/p/KUhyoJ9DiCa.

Reply:
`nil` is not a pointer, but a map is. When labels are `nil` and we initialize them, it does not have an effect. When labels is not `nil` and we change it, we change it for everyone, because a map is a pointer. 

* https://play.golang.org/p/tCO7P-APIO1
* https://dave.cheney.net/2017/04/30/if-a-map-isnt-a-reference-variable-what-is-it

Reply:
TIL. I always thought map is a value holding a reference. Still there are no writes to the map so the CR remains unchanged. But I'll simply replace that with:

```go
		var v string
		if cr.Labels != nil {
			v = cr.Labels[key.LabelReleaseCyclePhase]
		}
```

Reply:
> We could even only give a copy to all resources in `operatorkit` by default.

What do you think about that?

Reply:
I think it's a good idea. This way we are guaranteed we don't mess with cached objects.

Reply:
I made a note in my board.